### DISCUSSION TITLE
How to adjust the config for the network errors retry mechanism?

### DISCUSSION THREAD

**Discussion by ovidyou** [jsforce/jsforce#1605] [Q&A]
Hello! From the documentation is not clear how to configure the retry mechanism that is now integrated into version 3.
Can someone offer an example? Thank you!

**Comment by cristiand391** [comment]
Hey, the only way to configure retry options is by using the `request` helper directly:
https://github.com/jsforce/jsforce/blob/3de28423dfa71134509e907d86db2458d493dae0/src/request.ts#L231

opts:
https://github.com/jsforce/jsforce/blob/main/src/request.ts#L46-L69

You can't configure it globally for jsforce, this decision was on purpose because we don't know if we'll be able to support the network retry mechanism in the browser in the future (v3 uses node-fetch v2 for node & browsers, we need to drop it for the next major jsforce release).

**Reply by lucabezerra** [reply]
> we don't know if we'll be able to support the network retry mechanism in the browser in the future (v3 uses node-fetch v2 for node & browsers, we need to drop it for the next major jsforce release)

@cristiand391 I'm a bit confused by this part of your answer. The retries mechanism was just implemented as part of v3, did I get it right in understanding that you don't know if it'll be available on v4+? And more specifically, is this a concern only for the browser version of jsforce or also for the node version?
Thanks in advance!

**Reply by cristiand391** [reply]
@lucabezerra sorry for the late reply.
The retry mechanism will always work for jsforce-node, the tricky part is the browser implementation because the native `fetch` isn't as extensible as `node-fetch`.

**Comment by ovidyou** [comment]
Hi @cristiand391, thank you for your response.

We have several projects that rely on the `apex()` and `sobject()` wrappers. Refactoring everything to use `request()` would be quite a large effort. We would like to migrate to version 3, but currently are blocked by the fact that we cannot override the retry mechanism options.

I understand the retry mechanism might be dropped in the future. What do you think about adding a method to the connection object to override the retry options? In a future version, if the retry feature is removed, this method could simply log a console warning to inform users about the change.

Maybe something like below. Me and my team would really appreciate this feature!

```js
const jsforce = require('jsforce');
const conn = new jsforce.Connection();

conn.setRequestRetryOptions({
    maxRetries: 2,
    minTimeout: 1000,
    timeoutFactor: 3,
});
```

**Comment by tehnrd** [comment]
I created this [issue](https://github.com/jsforce/jsforce/issues/1643) before realizing there was this discussion. I would really like to try to fix this but I'm struggling to understand why this is a challenge to configure. Can't we set some options on the connection object and pass them to the request?
